addons/mail/models/mail_link_preview.py

```diff
@@ -0,0 +1,135 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models, fields
+
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+from lxml import html, etree
+from urllib.parse import urlparse
+import requests
+
+
+class LinkPreview(models.Model):
+    _name = 'mail.link.preview'
+    _description = "Store link preview data"
+
+    message_id = fields.Many2one('mail.message', string='Message', index=True, ondelete='cascade')
+    guest_id = fields.Many2one('mail.guest', string="Guest")
+    og_type = fields.Char('type')
+    source_url = fields.Char('url', required=True)
+    og_title = fields.Char('title')
+    og_image = fields.Char('image')
+    og_description = fields.Text('description')
+    og_mimetype = fields.Char('mimetype')
+
+    # Some website are blocking non browser user agent. We make ourself like
+    # firefox to avoid issue.
+    request_headers = {
+        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; rv:91.0) Gecko/20100101 Firefox/91.0'
+    }
+
+    def _link_preview_format(self):
```

Comment:
I would re-order the methods in this file so that crud operations are above, followed by fetch stuff, and format last